src/persist-client/src/internal/apply.rs

```diff
@@ -0,0 +1,436 @@
+            ));
+        }
+        if !PartialOrder::less_equal(&self.upper, state.upper()) {
+            return Err(format!(
+                "upper unexpectedly not leq {:?} vs {:?}",
+                self.upper.elements(),
+                state.upper().elements()
+            ));
+        }
+        Ok(())
+    }
+
+    fn update<K, V, D>(&mut self, state: &State<K, V, T, D>) {
+        debug_assert_eq!(self.validate_cached_version_of(state), Ok(()));
+        self.seqno = state.seqno;
+        if &self.upper != state.upper() {
+            self.upper.clone_from(state.upper());
+        }
+        if &self.since != state.since() {
+            self.since.clone_from(state.since());
+        }
+    }
+}
+
+impl<K, V, T, D> Applier<K, V, T, D>
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    pub(crate) async fn new(
+        cfg: PersistConfig,
+        shard_id: ShardId,
+        metrics: Arc<Metrics>,
+        state_versions: Arc<StateVersions>,
+        shared_states: &StateCache,
+    ) -> Result<Self, Box<CodecMismatch>> {
+        let shard_metrics = metrics.shards.shard(&shard_id);
+        let state = shared_states
+            .get(shard_id, || {
+                metrics
+                    .cmds
+                    .init_state
+                    .run_cmd(&shard_metrics, |_cas_mismatch_metric| {
+                        // No cas_mismatch retries because we just use the returned
+                        // state on a mismatch.
+                        state_versions.maybe_init_shard(&shard_metrics)
+                    })
+            })
+            .await?;
+        let cached_state = {
+            let state = state.lock().await;
+            CachedState {
+                shard_id,
+                seqno: state.seqno,
+                upper: state.upper().clone(),
+                since: state.since().clone(),
+            }
+        };
+        Ok(Applier {
+            cfg,
+            metrics,
+            shard_metrics,
+            state_versions,
+            state,
+            cached_state,
+        })
+    }
+
+    pub fn cached_state(&self) -> &CachedState<T> {
+        &self.cached_state
+    }
+
+    pub fn is_tombstone(&self) -> bool {
+        self.cached_state.upper.is_empty() && self.cached_state.since.is_empty()
+    }
+
+    #[cfg(test)]
+    pub async fn seqno_since(&self) -> SeqNo {
+        self.state.lock().await.seqno_since()
+    }
+
+    pub async fn all_fueled_merge_reqs(&self) -> Vec<FueledMergeReq<T>> {
+        // WIP safe to serialize this on the mutex?
+        self.state
+            .lock()
+            .await
+            .collections
+            .trace
+            .all_fueled_merge_reqs()
+    }
+
+    pub async fn snapshot(
+        &self,
+        as_of: &Antichain<T>,
+    ) -> Result<Result<Vec<HollowBatch<T>>, Upper<T>>, Since<T>> {
+        // WIP safe to serialize this on the mutex?
```

Comment:
what's the definition of safety for the WIP ones here? safe in terms of correctness? or in terms of performance characteristics

Reply:
I can't think of why `snapshot` would be unsafe from either perspective, we get all of the batches from a single version of state in one go, and it should be quick to determine those batches

Reply:
oh, sorry. performance and availability. this e.g. would have to serialize with the commands wanting to lock the mutex